pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/statement/JsonStatementOptimizer.java

```diff
@@ -0,0 +1,561 @@
+              if (isIndexedJSONColumn(parts[0], config)) {
+                Function jsonMatchFunction = new Function("JSON_MATCH");
+
+                List<Expression> jsonMatchFunctionOperands = new ArrayList<>();
+                jsonMatchFunctionOperands.add(RequestUtils.createIdentifierExpression(parts[0]));
+                jsonMatchFunctionOperands.add(RequestUtils.createLiteralExpression(
+                    new StringLiteralAstNode(getJsonPath(parts, true) + getOperatorSQL(kind))));
+                jsonMatchFunction.setOperands(jsonMatchFunctionOperands);
+
+                expression.setFunctionCall(jsonMatchFunction);
+              } else {
+                operand.clear();
+                operand.setType(ExpressionType.FUNCTION);
+                operand.setFunctionCall(getJsonExtractFunction(parts, DataSchema.ColumnDataType.JSON));
+              }
+            }
+          }
+          break;
+        }
+      }
+    }
+  }
+
+  /**
+   *  @return A string array containing all the parts of an identifier. An identifier may have one or more parts that
+   *  are joined together using <DOT>. For example the identifier "testTable.jsonColumn.name.first" consists up of
+   *  "testTable" (name of table), "jsonColumn" (name of column), "name" (json path), and "first" (json path). The last
+   *  two parts when joined together (name.first) represent a JSON path expression.
+   */
+  private static String[] getIdentifierParts(Identifier identifier) {
+    return identifier.getName().split("\\.");
+  }
+
+  /**
+   * Builds a json path expression when given identifier parts. For example,given [jsonColumn, name, first], this
+   * function will return "$.name.first" as json path expression.
+   * @param parts identifier parts
+   * @param applyDoubleQuote delimit json path with double quotes if true; otherwise, don't delimit json path.
+   * @return JSON path expression associated with the given identifier parts.
+   */
+  private static String getJsonPath(String[] parts, boolean applyDoubleQuote) {
+    StringBuilder builder = new StringBuilder();
+    if (applyDoubleQuote) {
+      builder.append("\"");
+    }
+
+    builder.append("$");
+    for (int i = 1; i < parts.length; i++) {
+      builder.append(".").append(parts[i]);
+    }
+
+    if (applyDoubleQuote) {
+      builder.append("\"");
+    }
+
+    return builder.toString();
+  }
+
+  /** @return true if specified column has column datatype of JSON; otherwise, return false */
+  private static boolean isValidJSONColumn(String columnName, @Nullable Schema schema) {
+    return schema != null && schema.hasColumn(columnName) && schema.getFieldSpecFor(columnName).getDataType()
+        .equals(FieldSpec.DataType.JSON);
+  }
+
+  /** @return true if specified column has a JSON Index. */
+  private static boolean isIndexedJSONColumn(String columnName, @Nullable TableConfig config) {
+    return config != null && config.getIndexingConfig().getJsonIndexColumns().contains(columnName);
```

Comment:
`config.getIndexingConfig().getJsonIndexColumns()` might be `null` which will cause NPE

Reply:
Fixed.